### PR TITLE
8286897: Loom: Cleanup x86_64 StubGenerator

### PR DESCRIPTION
**Pull request by shipilev**
During x86_32 porting, I ended up rewriting the copy of x86_64 StubGenerator significantly to make it more streamlined. The original x86_64 StubGenerator should enjoy the similar streamlining, since it would likely to be the template for many other architectures. 

There should be no behavior changes, except one small one. For some weird reason, the `__ movptr(c_rarg1, rsp);` in `generate_cont_doYield` used to be much earlier, between `start` and `the_pc`. I don't believe that's part of the frame setup. It looks like a copy-paste glitch from `generate_throw_exception`. So I just moved it here. No tests complain about this move.

The rest of the changes themselves should speak for themselves, but here is a brief list:
 - no need for `assert_asm` macros, instead do the inline checks, like the rest of Hotspot and Loom code
 - `_masm` and `masm` shadowing can just use whatever is currently defined as `__`
 - more strong cohesion between `post_call_nop`, its associated PC and its oopmap
 - consistently use `call_VM_leaf` that only takes the number of arguments, as we are doing argument pushes ourselves anyway;
 - for many moves, it is easier to use `ptr` flavors (e.g. `testq` -> `testptr`) to ease copy-pasting to other arches;
 - `generate_cont_thaw` should just do the `enabled()` check and carry the `StubCodeMark`, like other stubs do;
 - inline comments are reflowed to block-like comments, matching the usual style of stub code;
 - labels are prefixed with `L_` for clarity (used in some places in Hotspot);
 - `jfr_prolog` and `jfr_epilog` are used only once and do not carry their weight;
 - `continuation_*` are documented a bit better;

Additional testing:
 - [x] Linux x86_64 fastdebug, `java/lang/Thread/virtual`
 - [x] Linux x86_64 fastdebug, `serviceability/jvmti`
 - [x] Linux x86_64 fastdebug, `runtime/vthread`
 - [x] Linux x86_64 fastdebug, `tier1`
 - [x] Linux x86_64 fastdebug, `tier2`
 - [x] Linux x86_64 fastdebug, `tier3`

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (1 review required, with at least 1 reviewer)

### Issue
 * [JDK-8286897](https://bugs.openjdk.java.net/browse/JDK-8286897): Loom: Cleanup x86_64 StubGenerator

### Reviewers
 * [Ron Pressler](https://openjdk.java.net/census#rpressler) (@pron - Author) ⚠️ Review applies to [b4a29e72](https://git.openjdk.java.net/jdk/pull/8755/files/b4a29e72a67d5fb383f2afc61c7db185fd5ada44)
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/8755/head:pull/8755` \
`$ git checkout pull/8755`

Update a local copy of the PR: \
`$ git checkout pull/8755` \
`$ git pull https://git.openjdk.java.net/jdk pull/8755/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 8755`

View PR using the GUI difftool: \
`$ git pr show -t 8755`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/8755.diff">https://git.openjdk.java.net/jdk/pull/8755.diff</a>

</details>

### CODE REVIEW

**Comment by pron** [comment]
The changes looks fine, but why do you want to remove assert_asm? It's useful even though it's new and not applied to old code. But if there's a good reason to do so, then it should also be removed from the aarch64 port.

**Comment by shipilev** [comment]
> The changes looks fine, but why do you want to remove assert_asm? It's useful even though it's new and not applied to old code. 

It is quite fiddly and clunky, TBH. My concern is accidentally putting the wrong inline assembly into macro and producing weird side-effects that behave differently when assertions are enabled. This is a risk every assertion carries, the assembly carries lots of risk on its own. I don't even want to start thinking how would a badly matched macro play out. (Also, in x86_32 port, I discovered the definition is under wrong `LP64` block, which ends up with funky linking errors).

I can leave `assert_asm` in, but this cleanup had removed the only two uses.

> But if there's a good reason to do so, then it should also be removed from the aarch64 port.

That would be something to do when cleaning up AArch64 StubGenerator.

File: src/hotspot/cpu/x86/stubGenerator_x86_64.cpp [hunk 1]
```diff
@@ -7429,192 +7429,199 @@ address generate_avx_ghash_processBlocks() {
...
-    __ post_call_nop(); // this must be exactly after the pc value that is pushed into the frame info, we use this nop for fast CodeBlob lookup
+    int frame_complete = the_pc - start;
+
+    // This nop must be exactly at the PC we push into the frame info.
+    // We use this nop for fast CodeBlob lookup, associate the OopMap
+    // with it right away.
+    __ post_call_nop();
+    OopMapSet* oop_maps = new OopMapSet();
+    OopMap* map = new OopMap(framesize, 1);
+    oop_maps->add_gc_map(the_pc - start, map);
```
Comment (vnkozlov) [inline src/hotspot/cpu/x86/stubGenerator_x86_64.cpp:7455]: Use `frame_complete`.

File: src/hotspot/cpu/x86/stubGenerator_x86_64.cpp [hunk 2]
```diff
@@ -7429,192 +7429,199 @@ address generate_avx_ghash_processBlocks() {
...
 
-    __ movptr(c_rarg0, r15_thread);
     __ set_last_Java_frame(rsp, rbp, the_pc);
+    __ movptr(c_rarg0, r15_thread);
+    __ movptr(c_rarg1, rsp);
     __ call_VM_leaf(Continuation::freeze_entry(), 2);
```
Comment (vnkozlov) [inline src/hotspot/cpu/x86/stubGenerator_x86_64.cpp:7460]: Was it bug to not set `c_rarg1`?  stubGenerator_aarch64.cpp has the same issue.

**Review by vnkozlov — COMMENTED** [review]
Nice cleanup.

File: src/hotspot/cpu/x86/stubGenerator_x86_64.cpp (hunk 2 above)
Comment (shipilev) [reply src/hotspot/cpu/x86/stubGenerator_x86_64.cpp:7460]: Ah, so this one is little "behavioral" change. For some weird reason, the `__ movptr(c_rarg1, rsp);` used to be much earlier, between `start` and `the_pc`. I don't believe that's part of the frame setup. It looks like a copy-paste glitch from `generate_throw_exception`. So I just moved it here. No tests complain about this move.

File: src/hotspot/cpu/x86/stubGenerator_x86_64.cpp (hunk 1 above)
Comment (shipilev) [reply src/hotspot/cpu/x86/stubGenerator_x86_64.cpp:7455]: Done.

File: src/hotspot/cpu/x86/stubGenerator_x86_64.cpp [hunk 3]
```diff
@@ -7627,30 +7634,44 @@ address generate_avx_ghash_processBlocks() {
...
+    bs->load_at(_masm, ACCESS_READ | IN_NATIVE, T_OBJECT, rax, Address(rax, 0), c_rarg0, r15_thread);
+
+    __ bind(L_null_jobject);
+
     __ leave();
     __ ret(0);
 
+    OopMapSet* oop_maps = new OopMapSet();
     OopMap* map = new OopMap(framesize, 1); // rbp
     oop_maps->add_gc_map(the_pc - start, map);
```
[the remark this answers is not in the document]
Comment (shipilev) [reply src/hotspot/cpu/x86/stubGenerator_x86_64.cpp:7666]: Done.

File: src/hotspot/cpu/x86/stubGenerator_x86_64.cpp (hunk 2 above)
Comment (vnkozlov) [reply src/hotspot/cpu/x86/stubGenerator_x86_64.cpp:7460]: Okay. I missed it in original place (too far away) and I thought it is new instruction.

**Review by vnkozlov — APPROVED** [review]
Good.

**Comment by shipilev** [comment]
> Good.

Thanks! Any more comments/testing, anyone? I would like to integrate it soon.

**Comment by pron** [comment]
You generally need to test `jdk/internal/vm/Continuation/` for all continuation-related stuff, but those are in tier1, I think.

**Comment by shipilev** [comment]
> You generally need to test `jdk/internal/vm/Continuation/` for all continuation-related stuff, but those are in tier1, I think.

I did test groups from #8765 on Linux x86_64, and they still pass.

**Comment by shipilev** [comment]
I am integrating now to provide cleaner base for porting.

/integrate

**Pull request closed without merging**